backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/tutor/user/UserService.java

```diff
@@ -153,4 +168,149 @@ public void resetDemoStudents() {
+    public void importListOfUsers(InputStream stream, int courseExecutionId) {
+        String line = "";
+        String cvsSplitBy = ",";
+        User.Role auxRole;
+        InputStreamReader isr = new InputStreamReader(stream, StandardCharsets.UTF_8);
+        try (BufferedReader br = new BufferedReader(isr)) {
+            while ((line = br.readLine()) != null) {
+                String[] userInfo = line.split(cvsSplitBy);
+                if (userInfo.length == 2) {
+                    auxRole = User.Role.STUDENT;
+                }
+                else if (userInfo.length > 2) {
+                    if (userInfo[2].equals("STUDENT")) {
+                        auxRole = User.Role.STUDENT;
+                    }
+                    else {
+                        auxRole = User.Role.TEACHER;
+                    }
+                }
+                else {
+                    throw new TutorException(INVALID_CSV_FILE_FORMAT);
+                }
+                ExternalUserDto userDto = new ExternalUserDto();
+                userDto.setEmail(userInfo[0]);
+                userDto.setName(userInfo[1]);
+                userDto.setRole(auxRole);
+                createExternalUser(courseExecutionId, userDto);
+            }
+        } catch (IOException ex) {
+            throw new TutorException(ErrorMessage.CANNOT_OPEN_FILE);
+        }
+    }
+
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public ExternalUserDto createExternalUser(Integer courseExecutionId, ExternalUserDto externalUserDto) {
+        verifyEmail(externalUserDto);
+        verifyRole(externalUserDto);
+        CourseExecution courseExecution = getCourseExecution(courseExecutionId);
+        User user1 = getUser(externalUserDto, courseExecution);
+        associateUserWithExecution(courseExecution, user1);
+        sendConfirmationEmailTo(user1);
+        return new ExternalUserDto(user1);
+    }
+
+    public String generateConfirmationToken(User user) {
+        String token = KeyGenerators.string().generateKey();
+        user.setTokenGenerationDate(LocalDateTime.now());
+        user.setConfirmationToken(token);
+        return token;
+    }
+
+    private void sendConfirmationEmailTo(User user) {
+        generateConfirmationToken(user);
+        mailer.sendSimpleMail(Mailer.MAIL_USER, user.getEmail(), User.PASSWORD_CONFIRMATION_MAIL_SUBJECT, buildMailBody(user));
+    }
+
+    private String buildMailBody(User user) {
+        String msg = "To confirm your registration click the following link";
+        return String.format("%s: %s", msg, LinkHandler.createConfirmRegistrationLink(user.getEmail(), user.getConfirmationToken()));
+    }
+
+    private void associateUserWithExecution(CourseExecution courseExecution, User user1) {
+        courseExecution.addUser(user1);
+        user1.addCourse(courseExecution);
+    }
+
+    private User getUser(ExternalUserDto externalUserDto, CourseExecution courseExecution) {
```

Comment:
The method name is misleading... associates an existing user with course execution or creates a new user (? without name?)